examples/Auction/StateMachine.hs

```diff
@@ -0,0 +1,339 @@
+module Auction.StateMachine (
+  auctionTests,
+  auctionTest,
+) where
+
+import Hedgehog (
+  Callback (Ensure, Require, Update),
+  Command (Command),
+  Concrete (Concrete),
+  FunctorB,
+  Group (Group),
+  MonadGen,
+  Property,
+  PropertyT,
+  Rec (Rec),
+  Symbolic,
+  TraversableB,
+  Var (Var),
+  annotateShow,
+  concrete,
+  executeSequential,
+  failure,
+  forAll,
+  property,
+  (===),
+ )
+import Hedgehog.Internal.Distributive (MonadTransDistributive (distributeT))
+
+import Control.Arrow (second)
+import Control.Monad (guard, liftM2)
+import Control.Monad.State.Strict (StateT, evalStateT, runState, state)
+import Data.Function ((&))
+import Data.Kind (Type)
+import Data.List (sort)
+import Data.Map (Map)
+import Data.Maybe (fromMaybe, isJust, isNothing)
+import GHC.Generics (Generic)
+import Plutus.Model (
+  Mock,
+  Run (Run),
+  adaValue,
+  checkErrors,
+  defaultBabbage,
+  initMock,
+ )
+import PlutusLedgerApi.V1 (PubKeyHash)
+import PlutusLedgerApi.V1.Contexts (TxOutRef)
+
+import Auction.PSM qualified as PSM
+import Data.Map qualified as Map
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Internal.Range qualified as Range
+
+newtype User = User {name :: String}
+  deriving stock (Eq, Ord, Show)
+
+data AuctionState v = AuctionState
+  { currentBid :: Maybe (User, User, Int, Var TxOutRef v)
+  , -- auction owner, current bidder, bid, current utxo
+    winner :: Maybe User
+  , users :: Map User (Int, Var PubKeyHash v)
+  }
+  deriving stock (Eq, Ord, Show)
+
+initialState :: AuctionState v
+initialState =
+  AuctionState
+    { currentBid = Nothing
+    , winner = Nothing
+    , users = Map.empty
+    }
+
+data AddUser v
+  = AddUser User Int
+  deriving stock (Eq, Show, Generic)
```

Comment:
I don't think we lose any power by requiring all users to be set up at mockchain creation and holding users immutable throught testing.

Reply:
Yeah I don't think there's any benefit to doing it this way, it was just easier.